### PR TITLE
Add nightly job to track build metrics

### PR DESCRIPTION
**Pull request by assignUser**
This PR adds a nightly job that collect build time and binary size metrics for a debug and release build and uploads the data to conbench. For example: https://velox-conbench.voltrondata.run/runs/BM-release-8665191090-1/

The adapters also add aggregate statistics, for time the `wall_time` is of course dependent on the number of cores so there will be clear differences between release and debug as debug runs on 16 cores (because the 8 cores don't have enough disk space for a debug build).

### CODE REVIEW

File: scripts/benchmark-requirements.txt [hunk 1]
```diff
@@ -12,6 +12,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-benchadapt@git+https://github.com/conbench/conbench.git@44e81d1#subdirectory=benchadapt/python
-benchalerts@git+https://github.com/conbench/conbench.git@44e81d1#subdirectory=benchalerts
-benchclients@git+https://github.com/conbench/conbench.git@44e81d1#subdirectory=benchclients/python
+benchadapt
+benchalerts
+benchclients
```
Comment (austin3dickey) [inline scripts/benchmark-requirements.txt:17]: cool, mental note to check on these builds after our releases to PyPI!

**Review by austin3dickey — APPROVED** [review]
looks great!

File: scripts/benchmark-requirements.txt (hunk 1 above)
Comment (assignUser) [reply scripts/benchmark-requirements.txt:17]: We can also pin to the current version to be safe but the previously pinned version was like a year old

**Review by austin3dickey — APPROVED** [review]
LGTM!

File: scripts/build-metrics.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+                    "data": [duration],
+                    "unit": "s",
+                    "iterations": 1,
+                },
+                tags={
+                    "name": object_path,
+                    "suite": suite,
+                    "source": f"{self.build_type}_build_metrics_time",
+                },
+                info={},
```
Comment (kgpai) [inline scripts/build-metrics.py:176]: i wonder if we can capture other things like ' compiler' , compiler flags used etc ? (maybe in the next set of changes).

File: scripts/build-metrics.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    sizes = BinarySizeAdapter(
+        command=["true"],
+        size_file=join(args.base_path, args.size_file),
+        build_type=args.build_type,
+        result_fields_override={
+            "run_id": args.run_id,
+            "run_name": run_name,
+            "run_reason": run_reason,
+            "github": {
+                # TODO change
```
Comment (kgpai) [inline scripts/build-metrics.py:221]: nit: what should be changed, can you add a more detailed comment.

**Review by kgpai — COMMENTED** [review]
Thanks ! This is very exciting, some small nits, comments.

File: .github/workflows/build-metrics.yml [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  metrics:
+    name: Linux ${{ matrix.type }} with adapters
+    if: ${{ github.repository == 'facebookincubator/velox' }}
+    runs-on: ${{ matrix.runner }}
+    container: ghcr.io/facebookincubator/velox-dev:adapters
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - runner: "8-core"
```
[the remark this answers is not in the document]
Comment (assignUser) [reply .github/workflows/build-metrics.yml:44]: Hm, does it even make sense to compare across build types? Also I think the only actual runner-caused difference should be in wall_time as the single objects are not build across cores. (well and maybe total linking time?) But the difference inherent in debug vs release are likely dominating that no?

(We can of course do that but my idea was that we will switch the debug build to 8 core as soon as we optimized the build not to need 70+GB  xD)

File: .github/workflows/build-metrics.yml [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+          make '${{ matrix.type }}'
+
+      - name: Log binary sizes
+        run: |
+          mkdir -p /tmp/metrics
+          sizes_file=/tmp/metrics/object_sizes
+          pushd '_build/${{ matrix.type }}'
+
+          find velox -type f -name '*.so' -o -name '*.a' -exec ls -l -BB {} \; |
+            awk '{print $5, $9; total += $5} END {print total," total_lib_size"}' > $sizes_file
```
[the remark this answers is not in the document]
Comment (assignUser) [reply .github/workflows/build-metrics.yml:89]: Yes, it creates a row with: <size> <name> for each binary(.so, .a and executables) and also adds a total sum for each group as the last row.

File: scripts/build-metrics.py (hunk 2 above)
Comment (assignUser) [reply scripts/build-metrics.py:176]: We can add arbitrary metadata as tags. There might also be a build in way to capture compiler and such similar to the machine info that's already part of the results.

File: scripts/build-metrics.py (hunk 3 above)
Comment (assignUser) [reply scripts/build-metrics.py:221]: ah, ups ^^ (it had my fork url before)

File: .github/workflows/build-metrics.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (kgpai) [reply .github/workflows/build-metrics.yml:45]: It does make some sense to compare across build types and flags - they are just compiler flags after all , but those comparisons will make less sense if the underlying hardware is not similar. When we switch over we can switch them both over, so its not really wasteful to ensure they are both on the same hardware i feel.

[the remark this answers is not in the document]
Comment (kgpai) [reply .github/workflows/build-metrics.yml:45]: For e.g I am especially curious to compare how much extra build time goes in optimization of some templated functions in debug to release mode etc.

[the remark this answers is not in the document]
Comment (assignUser) [reply .github/workflows/build-metrics.yml:45]: > e.g I am especially curious to compare how much extra build time goes in optimization of some templated functions in debug to release mode etc.

AH good point, that is interesting.

**Comment by assignUser** [comment]
I see the same error on main: https://github.com/facebookincubator/velox/actions/runs/8797715111/job/24143218933#step:8:3300

**Review by kgpai — APPROVED** [review]

**Pull request closed without merging**